File: BPMOPMobile_VietNamAirLines/BPMOPMobile_VietNamAirLines/BPMOPMobile.Droid/Presenter/SharedView/SharedView_PopupControlMultiChoice.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.Graphics.Drawables;
using Android.OS;
using Android.Runtime;
using Android.Support.V7.App;
using Android.Support.V7.Widget;
using Android.Views;
using Android.Widget;
using BPMOPMobile.Bean;
using BPMOPMobile.Droid.Class;
using BPMOPMobile.Droid.Core.Common;
using BPMOPMobile.Droid.Presenter.Adapter;
using BPMOPMobile.Droid.Presenter.Fragment;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace BPMOPMobile.Droid.Presenter.SharedView
{
    class SharedView_PopupControlMultiChoice : SharedView_PopupControlBase
    {
        public SharedView_PopupControlMultiChoice(LayoutInflater _inflater, AppCompatActivity _mainAct, CustomBaseFragment _fragment, string _fragmentTag, View _rootView) : base(_inflater, _mainAct, _fragment, _fragmentTag, _rootView)
        {

        }

        public override void InitializeValue_Master(ViewElement clickedElement)
        {
            base.InitializeValue_Master(clickedElement);
        }

        public override void InitializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JObject JObjectChild)
        {
            base.InitializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
        }

        public override void InitializeView()
        {
            base.InitializeView();
            try
            {
                #region Get View - Init Data

                View _viewPopupControl = _inflater.Inflate(Resource.Layout.PopupControl_SingleChoice, null);
                ImageView _imgClose = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_SingleChoice_Close);
                ImageView _imgDelete = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_SingleChoice_Delete);
                TextView _tvTitle = _viewPopupControl.FindViewById<TextView>(Resource.Id.tv_PopupControl_SingleChoice_Title);
                RecyclerView _recyData = _viewPopupControl.FindViewById<RecyclerView>(Resource.Id.recy_PopupControl_SingleChoice_Data);
                ImageView _imgDone = _viewPopupControl.FindViewById<ImageView>(Resource.Id.img_PopupControl_SingleChoice_Done);

                List<BeanLookupData> _lstLookupData = new List<BeanLookupData>();

                if (elementParent.Enable)
                {
                    _imgDone.Visibility = ViewStates.Visible;
                    _imgDelete.Visibility = ViewStates.Visible;
                }
                else
                {
                    _imgDone.Visibility = ViewStates.Invisible;
                    _imgDelete.Visibility = ViewStates.Invisible;
                }

                switch (base.flagView)
                {
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                    default:
                        {
                            _tvTitle.Text = elementParent.Title;
                            _lstLookupData = JsonConvert.DeserializeObject<List<BeanLookupData>>(elementParent.DataSource);

                            if (_lstLookupData != null && _lstLookupData.Count > 0 && !String.IsNullOrEmpty(elementParent.Value))
                            {
                                List<BeanLookupData> _lstValue = JsonConvert.DeserializeObject<List<BeanLookupData>>(elementParent.Value);

                                if (_lstValue == null || _lstValue.Count == 0)
                                    break;
                                foreach (var itemLookup in _lstLookupData)
                                {
                                    foreach (var itemValue in _lstValue)
                                    {
                                        if (itemLookup.ID == itemValue.ID && itemLookup.IsSelected == false) // item nào check qua rồi ko cần check lại.
                                            itemLookup.IsSelected = true;
                                    }
                                }
                            }
                            break;
                        }
                    case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                        {
                            _tvTitle.Text = elementPopup.Title;
                            _lstLookupData = JsonConvert.DeserializeObject<List<BeanLookupData>>(elementPopup.DataSource);

                            if (_lstLookupData != null && _lstLookupData.Count > 0 && !String.IsNullOrEmpty(elementPopup.Value))
                            {
                                List<BeanLookupData> _lstValue = JsonConvert.DeserializeObject<List<BeanLookupData>>(elementPopup.Value);

                                if (_lstValue == null || _lstValue.Count == 0)
                                    break;
                                foreach (var itemLookup in _lstLookupData)
                                {
                                    foreach (var itemValue in _lstValue)
                                    {
                                        if (itemLookup.ID == itemValue.ID && itemLookup.IsSelected == false) // item nào check qua rồi ko cần check lại.
                                            itemLookup.IsSelected = true;
                                    }
                                }
                            }
                            break;
                        }
                }

                AdapterFormControlMultiChoice _adapterFormControlMultiChoice = new AdapterFormControlMultiChoice((MainActivity)base._mainAct, _rootView.Context, _lstLookupData);
                StaggeredGridLayoutManager staggeredGridLayoutManager = new StaggeredGridLayoutManager(1, LinearLayoutManager.Vertical);
                _recyData.SetAdapter(_adapterFormControlMultiChoice);
                _recyData.SetLayoutManager(staggeredGridLayoutManager);
                #endregion

                #region Show View                
                Dialog _dialogPopupControl = new Dialog(_rootView.Context, Resource.Style.Theme_Custom_BPMOP_Dialog_FullScreen);
                Window window = _dialogPopupControl.Window;
                _dialogPopupControl.RequestWindowFeature(1);
                _dialogPopupControl.SetCanceledOnTouchOutside(false);
                _dialogPopupControl.SetCancelable(true);
                window.SetSoftInputMode(SoftInput.AdjustPan); // Để Hide soft keyboard ko bị Resize
                window.SetGravity(GravityFlags.Bottom);
                var dm = _mainAct.Resources.DisplayMetrics;

                _dialogPopupControl.SetContentView(_viewPopupControl);
                _dialogPopupControl.Show();
                WindowManagerLayoutParams s = window.Attributes;
                s.Width = WindowManagerLayoutParams.MatchParent;
                s.Height = WindowManagerLayoutParams.MatchParent;
                window.Attributes = s;
                window.SetBackgroundDrawable(new ColorDrawable(Color.Transparent));
                #endregion

                #region Event

                if (elementParent.Enable) // Nếu control dc Enable mới cho click
                {
                    _adapterFormControlMultiChoice.CustomItemClick += (sender, e) =>
                    {
                        for (int i = 0; i < _lstLookupData.Count; i++)
                        {
                            if (_lstLookupData[i].ID.Equals(e.ID))
                            {
                                _lstLookupData[i].IsSelected = !_lstLookupData[i].IsSelected;
                                _adapterFormControlMultiChoice.NotifyDataSetChanged();
                                break;
                            }
                        }
                    };
                }

                _imgClose.Click += (sender, e) =>
                {
                    _dialogPopupControl.Dismiss();
                };

                _imgDelete.Click += (sender, e) =>
                {
                    for (int i = 0; i < _lstLookupData.Count; i++)
                    {
                        _lstLookupData[i].IsSelected = false;
                    }
                    _imgDone.PerformClick();
                };

                _imgDone.Click += (sender, e) =>
                {
                    string _result = "";

                    List<BeanLookupData> _lstselected = _lstLookupData.Where(x => x.IsSelected == true).ToList(); // lấy ra những thằng đang chọn

                    if (_lstselected != null && _lstselected.Count > 0)
                    {
                        _result = JsonConvert.SerializeObject(_lstselected);
                    }

                    switch (base.flagView)
                    {
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow: // Control of Master
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForElement(elementParent, _result);
                                _frag._adapterDetailExpandControl.NotifyDataSetChanged();
                                break;
                            }
                        case (int)SharedView_PopupControlBase.FlagViewControlDynamic.DetailWorkflow_InputGridDetail:
                            {
                                FragmentDetailWorkflow _frag = (FragmentDetailWorkflow)_fragment;
                                _frag.UpdateValueForPopupGridDetail(elementParent, elementPopup, JObjectChild, _result);
                                break;
                            }
                    }
                    _dialogPopupControl.Dismiss();
                };

                #endregion
            }
            catch (Exception ex)
            {
#if DEBUG
                CmmDroidFunction.WriteTrackingError(this.GetType().Name, "InitializeView", ex);
#endif
            }
        }
    }
}

File: VNABPMOP/app/src/main/java/com/vuthao/bpmop/shareview/SharedView_PopupControlMultiChoice.java
package com.vuthao.bpmop.shareview;

import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.vuthao.bpmop.R;
import com.vuthao.bpmop.base.BroadcastUtility;
import com.vuthao.bpmop.base.Functions;
import com.vuthao.bpmop.base.model.custom.LookupData;
import com.vuthao.bpmop.base.model.dynamic.ViewElement;
import com.vuthao.bpmop.base.vars.VarsReceiver;
import com.vuthao.bpmop.core.Vars;
import com.vuthao.bpmop.shareview.adapter.FormControlMultiChoiceAdapter;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.stream.Collectors;

public class SharedView_PopupControlMultiChoice extends SharedView_PopupControlBase implements FormControlMultiChoiceAdapter.FormControlMultiChoiceListener, View.OnClickListener {
    private ArrayList<LookupData> lstLookupData = new ArrayList<LookupData>();
    private ArrayList<LookupData> lstselected = new ArrayList<LookupData>();
    private FormControlMultiChoiceAdapter adapter;
    private String result;

    private Dialog dialog;
    private View view;
    private ImageView imgClose;
    private ImageView imgDelete;
    private TextView tvTitle;
    private RecyclerView recyData;
    private ImageView imgDone;

    public SharedView_PopupControlMultiChoice(LayoutInflater inflater, Activity mainAct, String fragmentTag, View rootView) {
        super(inflater, mainAct, fragmentTag, rootView);
    }

    @Override
    public void initializeValue_Master(ViewElement viewElement) {
        super.initializeValue_Master(viewElement);
    }

    @Override
    public void initializeValue_InputGridDetail(ViewElement elementParent, ViewElement elementPopup, JSONObject JObjectChild) {
        super.initializeValue_InputGridDetail(elementParent, elementPopup, JObjectChild);
    }

    @Override
    public void initializeView() {
        super.initializeView();

        view = inflater.inflate(R.layout.popup_control_single_choice, null);
        imgClose = view.findViewById(R.id.img_PopupControl_SingleChoice_Close);
        imgDelete = view.findViewById(R.id.img_PopupControl_SingleChoice_Delete);
        tvTitle = view.findViewById(R.id.tv_PopupControl_SingleChoice_Title);
        recyData = view.findViewById(R.id.recy_PopupControl_SingleChoice_Data);
        imgDone = view.findViewById(R.id.img_PopupControl_SingleChoice_Done);

        if (elementParent.isEnable()) {
            imgDone.setVisibility(View.VISIBLE);
            imgDelete.setVisibility(View.VISIBLE);
        } else {
            imgDone.setVisibility(View.INVISIBLE);
            imgDelete.setVisibility(View.INVISIBLE);
        }

        ArrayList<LookupData> lstValue;
        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow:
            default: {
                tvTitle.setText(elementParent.getTitle());
                lstLookupData = new Gson().fromJson(elementParent.getDataSource(), new TypeToken<ArrayList<LookupData>>() {
                }.getType());

                if (!lstLookupData.isEmpty() && !Functions.isNullOrEmpty(elementParent.getValue())) {
                    lstValue = new Gson().fromJson(elementParent.getValue(), new TypeToken<ArrayList<LookupData>>() {
                    }.getType());

                    if (lstValue.isEmpty()) {
                        break;
                    }

                    for (LookupData itemLookup : lstLookupData) {
                        for (LookupData itemValue : lstValue) {
                            // item nào check qua rồi ko cần check lại.
                            if (itemLookup.getID().equals(itemValue.getID()) && !itemLookup.isSelected()) {
                                itemLookup.setSelected(true);
                            }
                        }
                    }
                }
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                tvTitle.setText(elementPopup.getTitle());
                lstLookupData = new Gson().fromJson(elementPopup.getDataSource(), new TypeToken<ArrayList<LookupData>>() {
                }.getType());

                if (!lstLookupData.isEmpty() && !Functions.isNullOrEmpty(elementPopup.getValue())) {
                    lstValue = new Gson().fromJson(elementPopup.getValue(), new TypeToken<ArrayList<LookupData>>() {
                    }.getType());

                    if (lstValue.isEmpty()) {
                        break;
                    }

                    for (LookupData itemLookup : lstLookupData) {
                        for (LookupData itemValue : lstValue) {
                            // item nào check qua rồi ko cần check lại.
                            if (itemLookup.getID().equals(itemValue.getID()) && !itemLookup.isSelected()) {
                                itemLookup.setSelected(true);
                            }
                        }
                    }
                }
                break;
            }
        }

        adapter = new FormControlMultiChoiceAdapter(mainAct, lstLookupData, this);
        StaggeredGridLayoutManager staggeredGridLayoutManager = new StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL);
        recyData.setAdapter(adapter);
        recyData.setLayoutManager(staggeredGridLayoutManager);

        // region Dialog
        dialog = new Dialog(mainAct, R.style.Theme_Custom_BPMOP_Dialog_FullScreen);
        Window window = dialog.getWindow();
        dialog.requestWindowFeature(1);
        dialog.setCanceledOnTouchOutside(false);
        dialog.setCancelable(true);
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN);
        window.setGravity(Gravity.BOTTOM);
        dialog.setContentView(view);
        dialog.show();

        WindowManager.LayoutParams s = window.getAttributes();
        s.width = WindowManager.LayoutParams.MATCH_PARENT;
        s.height = WindowManager.LayoutParams.MATCH_PARENT;
        window.setAttributes(s);
        // endregion

        imgDone.setOnClickListener(this);
        imgDelete.setOnClickListener(this);
        imgClose.setOnClickListener(this);
    }

    private void done() {
        lstselected = (ArrayList<LookupData>) lstLookupData.stream().filter(LookupData::isSelected).collect(Collectors.toList());
        if (!lstselected.isEmpty()) {
            result = new Gson().toJson(lstselected);
        }

        switch (flagView) {
            case Vars.FlagViewControlDynamic.DetailWorkflow: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATEFORM);
                intent.putExtra("element", new Gson().toJson(elementParent));
                intent.putExtra("newValue", result);
                BroadcastUtility.send(mainAct, intent);
                break;
            }
            case Vars.FlagViewControlDynamic.DetailWorkflow_InputGridDetail: {
                Intent intent = new Intent();
                intent.setAction(VarsReceiver.UPDATECHILDACTION);
                intent.putExtra("elementParent", new Gson().toJson(elementParent));
                intent.putExtra("elementChild", new Gson().toJson(elementPopup));
                intent.putExtra("json", JObjectChild.toString());
                intent.putExtra("newValue", result);
                BroadcastUtility.send(mainAct, intent);
                break;
            }
        }

        dialog.dismiss();
    }

    private void delete() {
        for (int i = 0; i < lstLookupData.size(); i++) {
            lstLookupData.get(i).setSelected(false);
        }
        imgDone.performClick();
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.img_PopupControl_SingleChoice_Done: {
                done();
                break;
            }
            case R.id.img_PopupControl_SingleChoice_Delete: {
                delete();
                break;
            }
            case R.id.img_PopupControl_SingleChoice_Close: {
                dialog.dismiss();
                break;
            }
        }
    }

    @Override
    public void OnClick(LookupData data) {
        //Nếu control dc Enable mới xử lý
        if (elementParent.isEnable()) {
            for (int i = 0; i < lstLookupData.size(); i++) {
                if (lstLookupData.get(i).getID().equals(data.getID())) {
                    lstLookupData.get(i).setSelected(!lstLookupData.get(i).isSelected());
                    adapter.notifyDataSetChanged();
                    break;
                }
            }
        }
    }
}
